Updated the strata levels and label in `visr()`

ddsjoberg opened the pull request:
Updated the strata levels and label in `visr()` for KM and cumulative incidence objects. Rather than displaying "strata" as the label, we now print the variable label (if available) or the variable name. The levels no longer include the hard-coded print of the variable name.

Example of the updated figure below!

closes #343

``` r
library(visR)
estimate_KM(data = adtte, strata ="TRTP") |> visr()
```

![](https://i.imgur.com/G4so0Lw.png)

<sup>Created on 2022-04-14 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

SHAESEN2 commented on R/add_CNSR.R:
```diff
@@ -103,7 +103,8 @@ add_CNSR.ggsurvfit <- function(gg, shape = 3, size = 2, ...){
                         shape = shape,
-                        size = size)
+                        size = size,
+                        show.legend = FALSE)
```
We decided not to include this on purpose. The coloring should be sufficient without showing the censoring symbols. If you have a specific business need, we can reopen the discussion.

ddsjoberg replied:
I am not sure I understand this comment. After my other updates, the CNSR symbol legend was being added rather than integrated with the line color legend. I agree the color legend should be sufficient, which is now what shows.

SHAESEN2 commented on R/tidyme.R:
```diff
@@ -87,9 +87,16 @@ tidyme.survfit <- function(x, ...) {
+
   attr(retme, "survfit_object") <- survfit_object
+  # modify strata label, removing ref to raw variable name
+  if (!is.null(purrr::pluck(survfit_object, "estimate_KM_args", "strata"))) {
```
can we avoid the dependency on purrr? I dont think we are using it somewhere else.

ddsjoberg replied:
gone!

SHAESEN2 commented on R/tidyme.R:
```diff
@@ -87,9 +87,16 @@ tidyme.survfit <- function(x, ...) {
+  # modify strata label, removing ref to raw variable name
+  if (!is.null(purrr::pluck(survfit_object, "estimate_KM_args", "strata"))) {
+    for (stratum in purrr::pluck(survfit_object, "estimate_KM_args", "strata")) {
+      retme[["strata"]] <- stringr::str_remove(retme[["strata"]], stringr::fixed(paste0(stratum, "=")))
```
We decided not to depend on stringr. BASE R should be sufficient for sting manipulation.

ddsjoberg replied:
gone!

SHAESEN2 commented on R/visr.R:
```diff
@@ -560,3 +562,11 @@ visr.tidycuminc <- function(x = NULL
 
+
+.construct_strata_label <- function(estimate_args, sep = ", ") {
+  if (is.null(estimate_args[["strata"]])) return(" ")
```
Do we expect this to be NULL?
If we estimate using the formulat ~ 1 we add an artifical stratum "Overall" to facilitate downsteam processing.

ddsjoberg replied:
i updated the structure, so this line no longer appears. In this setting, the strata arg was NULL

SHAESEN2 commented on R/estimate_KM.R:
```diff
@@ -72,13 +72,14 @@ estimate_KM <- function(
 
 # Capture input to validate user input for data argument -----------------
+  dots <- rlang::dots_list(...)
+  estimate_KM_args <- as.list(environment())
```
I would only keep what we actually need from the environment() call. In clinical trials we work with big data so we don't want to keep information that is not required eg estimate_KM_args[["data"]]. If the strata is NULL is should be defaulted to "Overall" with label "Overall"

SHAESEN2 replied:
I'm not sure if I see the point of this additional line since we know the strata and we can remove whatever is in front of 'xx = ' easily + look for the corresponding label in the dataframe. Perhaps another argument in survfit object that contains strata labels might be more useful?

SHAESEN2 replied:
What I was expecting here is that you are capturing the actual environment of 'data' so that we don't get in trouble with wrapper functions.

ddsjoberg replied:
now just exporting the strata variables as a named list that include the labels (if they are present). Add the env, I feel, should be addressed in another PR.

SHAESEN2 commented on R/estimate_KM.R:
```diff
@@ -72,13 +72,14 @@ estimate_KM <- function(
 ){
```
Can you document the added behaviour to the roxygen section? It would facilitate review as I am not quite sure about the implemented logic although I know the endgame.

SHAESEN2 replied:
Once the logic is agreed upon, we need to add UAT for the PR. I can assist with this if required.

ddsjoberg replied:
I added a note about the strata list that is now returned with the object.

SHAESEN2 commented on R/estimate_KM.R:
```diff
@@ -146,8 +147,9 @@ estimate_KM <- function(
-  
-  survfit_object$call[[1]] <- quote(survival::survfit)
+
+  survfit_object[["estimate_KM_args"]] <- estimate_KM_args
```
adding the entire list of information seems redundant. We have all this information already in the survfit object.

ddsjoberg replied:
updated to just return the strata label

ddsjoberg commented:
Thanks for the review @SHAESEN2 ! I think the updates def make the changes better.
- there are already unit tests in place for assessing the correct levels in the `visr()` legend
- can you write a unit test for testing the legend title? I am not familiar with the internals of ggplot objects, and don't know how to  construct the test. I think 2 tests: one without strata and one with strata would be good. When there are no strata, the title is `" "`, otherwise, the legend title is the variable name/label

SHAESEN2 commented on R/estimate_KM.R:
```diff
@@ -11,6 +11,7 @@
 #'      \item{The cumulative hazard is estimated using the Nelson-Aalen estimator (ctype = 1): H.tilde = cumsum(x$n.event/x$n.risk).
 #'      The MLE (H.hat(t) = -log(S.hat(t))) can't be requested.}
 #'      \item{A two-sided pointwise 0.95 confidence interval is estimated using a log transformation (conf.type = "log").}
+#'      \item{A named list of strata (and their lables), if present}
```
This item list are things to be aware of, not the content of the survival object. I propose you rephrase this or explain it somewhere else.

ddsjoberg replied:
updated!

SHAESEN2 commented on R/estimate_cuminc.R:
```diff
@@ -38,6 +38,8 @@ estimate_cuminc <- function(data
   # check for installation of tidycmprsk package
   rlang::check_installed("tidycmprsk", version = "0.1.1")
+  dots <- rlang::dots_list(...)
+  estimate_cuminc_args <- as.list(environment())
```
I would implement the labels in a similar way here.

ddsjoberg replied:
this line was left over from the last round of review. I removed the line that exports this object with the result, but forgot to delete the line created the object

SHAESEN2 commented on R/visr.R:
```diff
@@ -560,3 +562,66 @@ visr.tidycuminc <- function(x = NULL
+
+    # get the original data frame
+    if (inherits(x, "survfit")) {
+      data <- as.list(x$call)[["data"]] %>% eval()
```
If we have the logic of strata labels in estimate_xx it would make it easier. You would do all the heavy lifting in a spot where all the information is available. Using this logic, you wont need to rebuild what you originally had.

The disadvantage is when we are passing a foreign survift object to visr(). The labels wont be adjusted.

Do we sacrifice efficiency or compatibility?

ddsjoberg replied:
I am not sure what happened...I thought I had deleted these lines. Anyway, they are now removed!

ddsjoberg commented:
Thanks @SHAESEN2 again for the review. I addressed all comments, and was able to add unit tests confirming the legend title is correct/as expected.

Can you take another look?

SHAESEN2 commented on R/estimate_KM.R:
```diff
@@ -5,6 +5,7 @@
 #'    argument are removed.
 #'    Alternatively, PARAM/PARAMCD can be used in the \code{strata} argument. \cr
 #'    The result is an object of class \code{survfit} which can be used in downstream functions and methods that rely on the \code{survfit} class.
+#'    When stratum are present, the returned survfit object is supplemented with the a named list of the stratum and associated label, if present.
```
Documentation is clear now: stratum -> strata

ddsjoberg replied:
updated.

SHAESEN2 commented on R/estimate_KM.R:
```diff
@@ -184,6 +185,15 @@ estimate_KM <- function(
+    survfit_object[["strata_lbls"]] <-
+      lapply(as.list(strata), function(x) attr(data[[x]], "label") %||% x) %>%
+      rlang::set_names(strata)
+  }
```
else if (is.null(strata){
  survfit_object[["strata_lbls"]] <- "Overall"
  names(survfit_object[["strata_lbls"]]) <- "Overall"
}

SHAESEN2 replied:
After some spot checks, I noticed we also have the required information in the names attribute of the strata element. I'm OK with adding that information into a new list element, but it should be done consistently (same logic for the attribute and the list element)

ddsjoberg replied:
> else if (is.null(strata){ survfit_object[["strata_lbls"]] <- "Overall" names(survfit_object[["strata_lbls"]]) <- "Overall" }

This is for the label, not the levels (which are already handled)

> After some spot checks, I noticed we also have the required information in the names attribute of the strata element. I'm OK with adding that information into a new list element, but it should be done consistently (same logic for the attribute and the list element)

That object has the raw levels of the stratifying variables, not the labels for a strata variable

SHAESEN2 replied:
I think they can still be aligned correct?
We need to have "Overall" in the legend when no strata are specified

SHAESEN2 replied:
> I think they can still be aligned correct? We need to have "Overall" in the legend when no strata are specified

SHAESEN2 replied:
Can you put a comment to explain the differences between the list element of labels and the attribute above in the code?

Also explain briefly why you would not add "Overall" here.

ddsjoberg replied:
more detail added in the comments

SHAESEN2 commented on R/tidyme.R:
```diff
@@ -87,9 +87,21 @@ tidyme.survfit <- function(x, ...) {
-  
+
   attr(retme, "survfit_object") <- survfit_object
+  strata <- .extract_strata_varlist(survfit_object)
```
where is this function defined? I couldnt find it

ddsjoberg replied:
In RStudio, you can find any function object by pressing `Ctrl+.`

SHAESEN2 replied:
I know. I meant that there was no function defined in the files :)

ddsjoberg replied:
moved to utils file

SHAESEN2 commented on tests/testthat/test-estimate_KM.R:
```diff
@@ -22,6 +22,7 @@
 #' T3.3 When no strata are specified, an artificial strata is created 'Overall'
 #' T3.4 When only 1 stratum is specified, the stratum names are added to the `names` attribute'
```
We already had strata labels in the names attribute of survfit_object[["strata"]]. Can we not leverage that logic instead of adding another list to the survfit_object? Sorry to notice it this late.

For me its to include another element, but it seems overkill.

ddsjoberg replied:
that object was not modified in this PR, and those are the strata levels

SHAESEN2 commented on tests/testthat/test-estimate_KM.R:
```diff
@@ -22,6 +22,7 @@
 #' T3.3 When no strata are specified, an artificial strata is created 'Overall'
 #' T3.4 When only 1 stratum is specified, the stratum names are added to the `names` attribute'
 #' T3.5 When more than 1 strata is specified, the stratum names are available in the `names` attribute
+#' T3.6 The strata label is constructed correctly
```
I'm not in favor of general descriptions. Can you explain what you mean with 'correctly'?
Perhaps: The strata labels are also available in a separate list element of the survfit object.

You also need to write a test to ensure:
- that this list element is added to the object.
- that the content is equal to the label of the corresponding variable in the data
- test if combination of labels is done as expected
- test if Overall is used if ~1 in formula

ddsjoberg replied:
All of these are tested already, although not each an explicit separate test. 
- that this list element is added to the object: already testing equality to a list object, there it must be a list.
- that the content is equal to the label of the corresponding variable in the data: this is the test that already appears
- that the content is equal to the names attribute of the strata element of the survfit object: does not apply, those are strata levels, not the strata labels
- test if combination of labels is done as expected: tested in the visr file, where the strata label is checked
- test if Overall is used if ~1 in formula: also already in the visr test file

SHAESEN2 replied:
@SHAESEN2 to look into this again in final review.

SHAESEN2 commented on tests/testthat/test-tidyme.R:
```diff
@@ -75,8 +75,9 @@ testthat::test_that("T3.2 The S3 method, associated with a `survfit` object, has
 
   survfit_object <- visR::estimate_KM(data = adtte, strata = "TRTA")
   survfit_object_tidy <- tidyme(survfit_object)
+  survfit_object$strata_lbls <- NULL
```
This is cheating :) You need to add the labels to surv_object_df to demonstrate that each list element is available in the final dataframe

ddsjoberg replied:
corrected

ddsjoberg commented:
@SHAESEN2 comments addressed. if you have additional comments, perhaps we can jump on a call to resolve quickly.

SHAESEN2 commented on R/estimate_KM.R:
```diff
@@ -184,6 +185,15 @@ estimate_KM <- function(
     }
   }
 
+  # add strata labels ----------------------------------------------------------
```
Should not be a new section, but rather a comment.

ddsjoberg replied:
if that is your preference, i'll remove

SHAESEN2 commented on R/tidyme.R:
```diff
@@ -22,8 +22,8 @@
 #'
-#' @return Tibble containing all list elements of the S3 object as columns. 
-#'   The column 'strata' is a factor to ensure that the strata are sorted 
+#' @return Tibble containing all list elements of the S3 object as columns.
```
Dataframe instead of tibble?

ddsjoberg replied:
i updated this text

SHAESEN2 commented on tests/testthat/test-estimate_CUMINC.R:
```diff
@@ -6,6 +6,8 @@
 #' T1.1 No errors `estimate_cuminc()`
 #' T1.2 No errors `estimate_cuminc()` sister functions
+#' T2. The legend follows the model strata labels
+#' T2.2 The color legend title is accurate.
```
accurate = represents the model strata label

ddsjoberg commented on tests/testthat/test-visr_survfit.R:
```diff
@@ -43,6 +43,7 @@
 #' T3.7 The `fun` argument is stored in the final object as attribute `fun`.
 #' T4. The legend follows the model strata labels
 #' T4.1 The legend follows the model strata labels
```
more detail added

ddsjoberg commented on tests/testthat/test-visr_survfit.R:
```diff
@@ -43,6 +43,7 @@
 #' T3.7 The `fun` argument is stored in the final object as attribute `fun`.
 #' T4. The legend follows the model strata labels
 #' T4.1 The legend follows the model strata labels
+#' T4.2 The color legend title is accurate.
```
more detail added

SHAESEN2 commented:
Amazing update and well tested!
